### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by jesusmb1995**
Hi, I work in meteomatics and I will help you with the code review.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+        let status = format!("{}", response.status());
+        println!("Status: {}", status);
+        println!("Headers:\n{:#?}", response.headers());
+
+        let body = response.text().await.unwrap();
+        println!("Body:\n{}", body);
+
+        assert_eq!(status, "200 OK");
+        assert_ne!(body, "");
```
Comment (jesusmb1995) [inline src/lib.rs]: Handling of http status and parsing of body should be inside `query_time_series`, the user of the API does not need that information.

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        let locations: Locations = LocationsBuilder::default()
+            .coordinates(coordinates)
+            .build()
+            .unwrap();
+
+        // Call endpoint
+        let response = meteomatics_connector
+            .query_time_series(utc_vdt, parameters, locations, None)
+            .await
+            .unwrap();
```
Comment (jesusmb1995) [inline src/lib.rs:238]: `query_times_series` should directly return a tuple containing the times and a matrix with float values, and not an HTTP response with text. Meaning that I expect something like `query_time_series -> Result<(Times, Matrix), reqwest::Error>` where `pub type Matrix = Vec<Vec<f64>>` and `Times = Vec<f64>`

Comment (jesusmb1995) [reply src/lib.rs:238]: The matrix should be something like `resultMatrix [time][parameter]`, the each row correspond to a time step and each column to a parameter. For that you will need to parse the csv on the body.

If you look at the C++ version you will see how it returns a matrix and a vector with times: https://github.com/meteomatics/cpp-connector-api/blob/master/src/meteomatics_main.cpp

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            },
+            P {
+                k: "precip_1h",
+                v: Some("mm"),
+            },
+        ]);
+        let parameters: Parameters = ParametersBuilder::default()
+            .p_values(p_values)
+            .build()
+            .unwrap();
```
Comment (jesusmb1995) [inline src/lib.rs]: Why not have directly create `Vec<String> parameters`?  That would remove the need of a ParametersBuilder and keep it simpler and easier/shorter to write: `let parameters = !vec["t_2m:C".to_string(), "precip_1h:mm".to_string()]`

File: src/lib.rs (hunk 1 above)
Comment (jesusmb1995) [reply src/lib.rs]: Err() should be returned `if status != "200 OK"`

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            Opt {
+                k: "calibrated",
+                v: "true",
+            },
+        ]);
+        let optionals: Optionals = OptionalsBuilder::default()
+            .opt_values(opt_values)
+            .build()
+            .unwrap();
+
```
Comment (jesusmb1995) [inline src/lib.rs:113]: Too keep it simple and concise I think its better to allow to provide directly a `Vec<String>` such as: `let optionals = !vec["source=mix".to_string(), "calibrated=true".to_string()]` and later when building the URL possibly validate the strings from special characters.

File: src/lib.rs (hunk 3 above)
Comment (jesusmb1995) [reply src/lib.rs]: and later when building the URL possibly validate the strings from special characters.

**Comment by jesusmb1995** [comment]
Summary of new comments:
- The response object does not contain a separate 2D matrix as I would expect to easily manipulate the values. 
- The implementation is also missing a clear interface to create coordinates `from_grid` `from_point_list` `from_point` since right now it can be created from any vector of strings.

**Comment by uboznikov** [comment]
Could you please provide some concrete examples of each.

And also could you please list here all the functions / features expected, also with examples, and with reference to the Python SDK code, as from the original requirements this should be the same as with the Python SDK, per Christian.

**Comment by jesusmb1995** [comment]
> Could you please provide some concrete examples of each.
> 
> And also could you please list here all the functions / features expected, also with examples, and with reference to the Python SDK code, as from the original requirements this should be the same as with the Python SDK, per Christian.

- The response object does not contain a separate 2D matrix as I would expect to easily manipulate the values.

From `example.py` https://github.com/meteomatics/python-connector-api/blob/master/examples/example.py 
The returned `df_ts` is a complex object that has inside separate attributes for the values (2D array) and the index (date times). If you add a print below the call you can see `df_ts.index` contains a separate vector just for the datetimes and coordinate of a point, then `df_ts.values` is a standalone 2D indexable array that can be indexed as `df_ts.values[1][0]`. In your case `.values` and `.index` is missing from the `response_records`: it should be possible to have `response_records.values[4][5]` and `response_records.index[4]` (String)  instead of `response_records[4][1][5]`

```
        df_ts = api.query_time_series(coordinates_ts, startdate_ts, enddate_ts, interval_ts, parameters_ts,
                                      username, password, model, ens_select, interp_select,
                                      cluster_select=cluster_select)

        print(df_ts.index)
        print(df_ts.values)
        print(df_ts.values[1][0])
```

prints
```
MultiIndex([(47.249297, 9.342854, '2021-10-19 00:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 01:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 02:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 03:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 04:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 05:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 06:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 07:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 08:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 09:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 10:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 11:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 12:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 13:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 14:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 15:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 16:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 17:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 18:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 19:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 20:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 21:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 22:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-19 23:00:00+00:00'),
            (47.249297, 9.342854, '2021-10-20 00:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 00:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 01:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 02:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 03:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 04:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 05:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 06:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 07:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 08:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 09:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 10:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 11:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 12:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 13:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 14:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 15:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 16:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 17:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 18:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 19:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 20:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 21:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 22:00:00+00:00'),
            (     50.0,     10.0, '2021-10-19 23:00:00+00:00'),
            (     50.0,     10.0, '2021-10-20 00:00:00+00:00')],
           names=['lat', 'lon', 'validdate'])
[[ 1.1   0.  ]
 [ 1.4   0.  ]
 [ 0.9   0.  ]
 [ 2.7   0.  ]
 [ 1.8   0.  ]
 [ 1.    0.  ]
 [ 1.2   0.  ]
 [ 2.3   0.  ]
 [ 2.5   0.09]
 [ 3.1   0.05]
 [ 3.6   0.  ]
 [ 2.9   0.03]
 [ 3.4   0.06]
 [ 3.6   1.39]
 [ 3.8   0.91]
 [ 4.2   0.  ]
 [ 4.2   0.  ]
 [ 4.5   0.  ]
 [ 4.7   0.  ]
 [ 4.8   0.  ]
 [ 5.2   0.  ]
 [ 5.6   0.  ]
 [ 6.1   0.  ]
 [ 6.6   0.  ]
 [ 6.9   0.  ]
 [ 4.3   0.  ]
 [ 4.2   0.  ]
 [ 4.5   0.  ]
 [ 4.6   0.  ]
 [ 5.4   0.  ]
 [ 5.    0.11]
 [ 5.4   4.27]
 [ 7.3   4.1 ]
 [ 8.6   1.01]
 [ 9.1   1.88]
 [10.1   0.29]
 [11.1   0.06]
 [11.6   0.  ]
 [11.9   0.  ]
 [12.7   0.  ]
 [13.7   0.  ]
 [13.2   0.  ]
 [12.2   0.  ]
 [11.9   0.  ]
 [11.7   0.  ]
 [11.5   0.  ]
 [11.3   0.  ]
 [11.    0.  ]
 [10.7   0.  ]
 [10.5   0.  ]]
1.4
```

**Comment by chess-equality** [comment]
> In your case `.values` and `.index` is missing from the `response_records`: it should be possible to have `response_records.values[4][5]` and `response_records.index[4]` (String) instead of `response_records[4][1][5]`
> 



In fact they exist and are not missing.

https://github.com/chess-equality/rust-connector-api/blob/develop/src/entities/connector_response.rs#L10-L13

```
pub type ResponseHeader = Vec<String>;
pub type ResponseRecord = Vec<(String, Vec<f64>)>;
```

`response_headers` is `index`
`response_records` is `values`

For example:

`response_body.response_headers`
`response_body.response_records`

`response_body.response_headers[1]`
`response_body.response_records[1].0`
`response_body.response_records[1].1`
`...`

If you run the unit tests they will show these.

> names=['lat', 'lon', 'validdate'])

Box (lat / lon) is not yet implemented.

**Comment by jesusmb1995** [comment]
> ```
> pub type ResponseHeader = Vec<String>;
> pub type ResponseRecord = Vec<(String, Vec<f64>)>;
> ```
> 
> `response_headers` is `index` `response_records` is `values`
> 
No, `response_header` is not `index`, it would be equivalent to `df_ts.columns` in the python code wich prints `Index(['t_2m:C', 'rr_1h:mm'], dtype='object')` but that is different from `df_ts.index`.

The response record right now has the index together with the values `let (index, values) = row;`

```
>>>>>>>>>> ResponseRecords:
2021-10-19T17:15:44Z: 14.3,0
2021-10-19T18:15:44Z: 13.8,0
```
All the indexes (for row 0 [2021-10-19T17:15:44Z], for row 1 [2021-10-19T18:15:44Z], etc..) should be instead on its own separate vector.

**Comment by chess-equality** [comment]
> 
> No, `response_header` is not `index`, it would be equivalent to `df_ts.columns` in the python code wich prints `Index(['t_2m:C', 'rr_1h:mm'], dtype='object')` but that is different from `df_ts.index`.
> 
> The response record right now has the index together with the values `let (index, values) = row;`
> 
> ```
> >>>>>>>>>> ResponseRecords:
> 2021-10-19T17:15:44Z: 14.3,0
> 2021-10-19T18:15:44Z: 13.8,0
> ```
> 
> All the indexes (for row 0 [2021-10-19T17:15:44Z], for row 1 [2021-10-19T18:15:44Z], etc..) should be instead on its own separate vector.

Could you put here the URL of this example, like:

https://api.meteomatics.com/2021-10-20T00:00:00Z/t_2m:C/52.520551,13.461804/csv

Note that box / grid is not yet implemented.

Then from the result, list which should go to `index`, `values`, and `columns`, and their respective structures (vec, tuple, etc.).

Furnish complete examples, like in the documentation:

https://www.meteomatics.com/en/api/getting-started/
https://www.meteomatics.com/en/api/request/required-parameters/#time_description

> No, `response_header` is not `index`, it would be equivalent to `df_ts.columns`

So there is this requirement, `df_ts.columns`?

Also complete the requirements expected, see below.

> 
> And also could you please list here all the functions / features expected, also with examples, and with reference to the Python SDK code, as from the original requirements this should be the same as with the Python SDK, per Christian.
>

https://github.com/meteomatics/rust-connector-api/pull/1#issuecomment-941259886

**Comment by jesusmb1995** [comment]
> Could you put here the URL of this example, like:
> 
> https://api.meteomatics.com/2021-10-20T00:00:00Z/t_2m:C/52.520551,13.461804/csv
> 
> Note that box / grid is not yet implemented.
> 
> Then from the result, list which should go to `index`, `values`, and `columns`, and their respective structures (vec, tuple, etc.).

For the following URL taken from the same python SDK example (line 99 of examples/example.py) for the case it requests a single coordinate time series for two parameters
https://api.meteomatics.com/2021-11-04T00:00:00+00:00--2021-11-05T00:00:00+00:00:PT1H/t_2m:C,rr_1h:mm/47.249297,9.342854/html?connector=python_v2.2.1&model=mix&interp_select=gradient_interpolation

```
coordinates_ts = [(47.249297, 9.342854)]  # request single coordinate <---------------
...
        df_ts = api.query_time_series(coordinates_ts, startdate_ts, enddate_ts, interval_ts, parameters_ts,
                                      username, password, model, ens_select, interp_select,
                                      cluster_select=cluster_select)

        print(df_ts.columns)
        print(df_ts.index)
        print(df_ts.values)
```

With response

```
validdate;t_2m:C;rr_1h:mm
2021-11-04T00:00:00Z;-8.0;0.24
2021-11-04T01:00:00Z;-7.8;0.25
2021-11-04T02:00:00Z;-7.8;0.64
2021-11-04T03:00:00Z;-8.4;0.45
2021-11-04T04:00:00Z;-9.0;0.49
2021-11-04T05:00:00Z;-9.3;0.25
2021-11-04T06:00:00Z;-9.4;0.00
2021-11-04T07:00:00Z;-9.3;0.10
2021-11-04T08:00:00Z;-9.8;0.95
2021-11-04T09:00:00Z;-9.7;1.17
2021-11-04T10:00:00Z;-9.6;0.91
2021-11-04T11:00:00Z;-9.4;0.39
2021-11-04T12:00:00Z;-9.6;0.23
2021-11-04T13:00:00Z;-9.8;0.05
2021-11-04T14:00:00Z;-9.6;0.00
2021-11-04T15:00:00Z;-9.3;0.00
2021-11-04T16:00:00Z;-9.2;0.53
2021-11-04T17:00:00Z;-9.5;0.46
2021-11-04T18:00:00Z;-9.4;0.39
2021-11-04T19:00:00Z;-9.2;0.07
2021-11-04T20:00:00Z;-9.4;0.00
2021-11-04T21:00:00Z;-9.9;0.00
2021-11-04T22:00:00Z;-10.2;0.00
2021-11-04T23:00:00Z;-10.1;0.00
2021-11-05T00:00:00Z;-10.8;0.00
```
---

`df_ts.index` is a `Vector<String>`  with the date associated to each row: 

```
2021-11-04T00:00:00Z;
2021-11-04T01:00:00Z;
2021-11-04T02:00:00Z;
2021-11-04T03:00:00Z;
2021-11-04T04:00:00Z;
   ...
```

---
`df_ts.columns` is a `Vector<String>` with:  't_2m:C', 'rr_1h:mm'

> So there is this requirement, `df_ts.columns`?

Yes, see above

---

`df_ts.values` is a `Vector<Vector<f32>>` with 

```
[[ -8.     0.24]
 [ -7.8    0.25]
 [ -7.8    0.64]
 [ -8.4    0.45]
 [ -9.     0.49]
 [ -9.3    0.25]
 [ -9.4    0.  ]
 [ -9.3    0.1 ]
 [ -9.8    0.95]
 [ -9.7    1.17]
 [ -9.6    0.91]
 [ -9.4    0.37]
 [ -9.6    0.21]
 [ -9.8    0.02]
 [ -9.6    0.  ]
 [ -9.3    0.  ]
 [ -9.2    0.53]
 [ -9.5    0.46]
 [ -9.4    0.39]
 [ -9.2    0.07]
 [ -9.4    0.  ]
 [ -9.9    0.  ]
 [-10.2    0.  ]
 [-10.1    0.  ]
 [-10.8    0.  ]]
```

**Pull request closed without merging**